File: Python/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        visited = set()
        res = []
        self.permuteRecursive(nums, visited, [], res)
        return res
    def permuteRecursive(self, nums: List[int], visited: set, sub_set, res):
        if len(sub_set) >= len(nums):
            res.append(sub_set)
        for i in range(len(nums)):
            if i not in visited:
                visited.add(i)
                self.permuteRecursive(nums, visited, sub_set + [nums[i]], res)
                visited.remove(i)

File: Java/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        permuteRecursive(nums, new ArrayList<>(), res);
        return res;
    }
    public void permuteRecursive(int[] nums, List<Integer> tmpList, List<List<Integer>> res){
        if(tmpList.size() >= nums.length){
            res.add(new ArrayList<>(tmpList));
        }
        else{
            for(int i = 0; i < nums.length; i++){
                if(tmpList.contains(nums[i]))
                    continue;
                tmpList.add(nums[i]);
                permuteRecursive(nums, tmpList, res);
                tmpList.remove(tmpList.size() - 1);
            }
        }
    }
}